File: src/server/repositories/faculty.repository.js
import prisma from '../middleware/prisma.intercepter.js';

class FacultyRepository {
    async getAll({ page_number = 1, page_size = 10, search = '', is_deleted }) {
        try {
            console.log('is_deleted', is_deleted);

            const skip = (page_number - 1) * page_size;

            const [faculties, total] = await Promise.all([
                prisma.faculty.findMany({
                    where: {
                        is_deleted: is_deleted === 'false' ? false : is_deleted === 'true' ? true : null,
                        OR: [{ name: { contains: search } }],
                    },
                    include: {
                        _count: {
                            select: {
                                students: true,
                                lecturers: true,
                                classes: true,
                            },
                        },
                    },
                    skip,
                    take: parseInt(page_size),
                    orderBy: {
                        id: 'desc',
                    },
                }),

                prisma.faculty.count({
                    where: {
                        is_deleted: is_deleted === 'false' ? false : is_deleted === 'true' ? true : null,
                        OR: [{ name: { contains: search } }],
                    },
                }),
            ]);

            return {
                data: faculties,
                total_records: total,
                page_number,
                page_size,
            };
        } catch (error) {
            throw error;
        }
    }

    async getById(id) {
        try {
            return await prisma.faculty.findUnique({
                where: { id: parseInt(id) },
                include: {
                    students: {
                        include: {
                            user: true,
                        },
                    },
                    lecturers: {
                        include: {
                            user: true,
                        },
                    },
                    classes: true,
                    _count: {
                        select: {
                            students: true,
                            lecturers: true,
                            classes: true,
                        },
                    },
                },
            });
        } catch (error) {
            throw error;
        }
    }

    async create(data) {
        try {
            const existingFaculty = await prisma.faculty.findFirst({
                where: {
                    code: data.code,
                    name: data.name,
                },
            });

            if (existingFaculty) {
                throw new Error('Mã khoa đã tồn tại');
            }

            return await prisma.faculty.create({
                data: {
                    name: data.name,
                    code: data.code,
                },
            });
        } catch (error) {
            throw error;
        }
    }

    async update(id, data) {
        try {
            const existingFaculty = await prisma.faculty.findFirst({
                where: { id },
            });

            if (!existingFaculty) {
                throw new Error('Không tìm thấy khoa');
            }

            if (data.code && data.code !== existingFaculty.code) {
                const duplicateCode = await prisma.faculty.findUnique({
                    where: { code: data.code },
                });

                if (duplicateCode) {
                    throw new Error('Mã khoa đã tồn tại');
                }
            }

            return await prisma.faculty.update({
                where: { id },
                data: {
                    name: data.name,
                    code: data.code,
                    is_deleted: data?.is_deleted,
                },
            });
        } catch (error) {
            throw error;
        }
    }

    async delete(id) {
        try {
            const faculty = await prisma.faculty.findFirst({
                where: { id },
            });

            if (!faculty) {
                throw new Error('Không tìm thấy khoa');
            }

            await prisma.faculty.delete({
                where: { id },
            });

            return faculty;
        } catch (error) {
            throw error;
        }
    }

    async deleteSoft(id) {
        try {
            const faculty = await prisma.faculty.findFirst({
                where: { id },
            });

            if (!faculty) {
                throw new Error('Không tìm thấy khoa');
            }

            await prisma.faculty.update({
                where: { id },
                data: { is_deleted: true },
            });

            return {
                ...faculty,
                is_deleted: true,
            };
        } catch (error) {
            throw error;
        }
    }

    async restore(id) {
        try {
            const faculty = await prisma.faculty.findFirst({
                where: { id },
            });

            if (!faculty) {
                throw new Error('Không tìm thấy khoa');
            }

            await prisma.faculty.update({
                where: { id },
                data: { is_deleted: false },
            });

            return {
                ...faculty,
                is_deleted: false,
            };
        } catch (error) {
            throw error;
        }
    }
}

export default new FacultyRepository();

File: src/server/AspNet/Infrastructure/data/Repositories/FacultyRepository.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Domain.Entites;
using Domain.Interfaces;
using Infrastructure.data.context;
using Microsoft.EntityFrameworkCore;

namespace Infrastructure.data.Repositories
{
    public class FacultyRepository(ApplicationDbContext context) : IFacultyRepository
    {

        public async Task<Faculty> AddAsync(Faculty faculty)
        {
            var result = await context.Faculties.AddAsync(faculty);
            await context.SaveChangesAsync();
            return result.Entity;
        }

        public async Task<Faculty> DeleteAsync(Faculty faculty)
        {
            var result = context.Faculties.Remove(faculty).Entity;
            await context.SaveChangesAsync();
            return result;
        }
        public async Task<Faculty> DeleteSoftAsync(Faculty faculty)
        {
            faculty.isDeleted = true;
            var result = context.Faculties.Update(faculty);
            await context.SaveChangesAsync();
            return result.Entity;
        }

        public async Task<(IEnumerable<Faculty> Items, int totalRecords, int pageNumber, int pageSize)> GetAllAsync(
            int pageNumber = 1,
            int pageSize = 10,
            string? search = null,
            bool? isDeleted = null
        )
        {
            var query = context.Faculties.AsQueryable();

            if (!string.IsNullOrEmpty(search))
            {
                query = query.Where(f => f.name.Contains(search) || f.code.Contains(search));
            }

            if (isDeleted != null)
            {
                query = query.Where(f => f.isDeleted == isDeleted);
            }

            var totalCount = await query.CountAsync();

            if (totalCount == 0)
            {
                return (new List<Faculty>(), totalCount, pageNumber, pageSize);
            }

            var items = await query
                .Skip((pageNumber - 1) * pageSize)
                .Take(pageSize)
                .ToListAsync();

            return (items, totalCount, pageNumber, pageSize);
        }


        public Task<Faculty?> GetByCodeAsync(string code)
        {
            return context.Faculties.FirstOrDefaultAsync(f => f.code == code);
        }

        public Task<Faculty?> GetByIdAsync(string id)
        {
            return context.Faculties.FirstOrDefaultAsync(f => f.Id == id);
        }

        public Task<Faculty?> GetByNameAsync(string name)
        {
            return context.Faculties.FirstOrDefaultAsync(f => f.name == name);
        }

        public async Task<bool> HasDependentEntitiesAsync(string id)
        {
            var hasMajor = await context.Majors.AnyAsync(m => m.facultyId == id);
            var hasLecturer = await context.Lecturers.AnyAsync(l => l.facultyId == id);
            var hasStudent = await context.Students.AnyAsync(s => s.facultyId == id);
            return hasMajor || hasLecturer || hasStudent;
        }

        public async Task<Faculty> RestoreAsync(Faculty faculty)
        {
            faculty.isDeleted = false;
            var result = context.Faculties.Update(faculty);
            await context.SaveChangesAsync();
            return result.Entity;
        }

        public async Task<Faculty> UpdateAsync(Faculty faculty)
        {
            var result = context.Faculties.Update(faculty);
            await context.SaveChangesAsync();
            return result.Entity;
        }
    }
}